File: solutions/clojure/raindrops/src/raindrops.clj
(ns raindrops)

(defn convert
  "Convert a number into its corresponding raindrop sounds."
  [number]
  (letfn [(divisible-by? [d] (zero? (mod number d)))
          (factor [num, drop] (when (divisible-by? num) drop))]
    (let [sound (str (factor 3 "Pling") (factor 5 "Plang") (factor 7 "Plong"))]
      (if (empty? sound) (str number) sound))))

File: solutions/fsharp/raindrops/Raindrops.fs
module Raindrops

let convert (number: int) : string =
    [ (3, "Pling"); (5, "Plang"); (7, "Plong") ]
    |> List.choose (fun (factor, sound) -> if number % factor = 0 then Some sound else None)
    |> function
        | [] -> string number
        | sounds -> String.concat "" sounds
